src/helpers.coffee

```diff
@@ -111,12 +111,35 @@ buildLocationData = (first, last) ->
 # This returns a function which takes an object as a parameter, and if that
 # object is an AST node, updates that object's locationData.
 # The object is returned either way.
-exports.addLocationDataFn = (first, last) ->
+exports.addDataToNode = (parserState, first, last) ->
   (obj) ->
-    if ((typeof obj) is 'object') and (!!obj['updateLocationDataIfMissing'])
+    # Add location data
+    if obj?.updateLocationDataIfMissing? and first?
       obj.updateLocationDataIfMissing buildLocationData(first, last)
 
-    return obj
+    # Add comments data
+    unless parserState.tokenComments
```

Comment:
Nit: might not help, but `parserState.tokenComments` is repeated a lot in this block. It could be a bit more readable if we pull that out first:

```coffee
{ tokenComments } = parserState
…
unless tokenComments[tokenHash]?
  tokenComments[tokenHash] = token.comments
else
  tokenComments[tokenHash] = tokenComments[tokenHash].concat token.comments
…
```